text/0036-consensus-memory-usage.md

```diff
@@ -0,0 +1,130 @@
+
+
+### Finality is determined by finality signatures only
+
+Once all nodes in the network have dropped the protocol state in an era, an attacker controlling a majority of the stake could now create an alternative protocol state: The equivocation could not be proven anymore, because everyone forgot about the previous instance.
+
+That's why a node should only consider a block to be truly finalized once it has enough finality signatures. The in-protocol finalization based on Highway summits is only used as a trigger to create a finality signature.
+
+That way, if the attacker wants to make us believe in their fork, they will have to create finality signatures, too, and for those they will still get slashed. If they don't produce those, the user won't see any finalized blocks and realize that they are not synced and connected to the live network.
+
+
+### Emergency restarts
+
+In the event of equivocation catastrophes, honest validators may end up signing conflicting forks. So the rule that the parent must also be signed does not apply here. After an emergency restart, you _are_ allowed to sign a different fork than before the restart, and cannot get slashed for it.
+
+
+## Drawbacks
+
+[drawbacks]: #drawbacks
+
+TBD
+
+
+## Rationale and alternatives
+
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+Slashing for finality signatures needs to be implemented anyway at some point, and the rest of the solution is relatively simple and doesn't even use any disk space.
+
+
+### Alternative: Swapping out whole eras to disk
+
+Most old eras are inactive and don't change anymore, because nobody is creating units for them. We could swap them out to disk and just enqueue any incoming messages for them, and only swap them in (at most one at a time) when enough messages are waiting.
+
+However, this CEP argues that it's not necessary to store them at all, which is simpler and less wasteful.
+
+
+### Alternative: Backing protocol state by a database entirely
+
+We could replace the `units` and `blocks` collections in `highway::State` with something backed by a database, that transparently caches the entries. In other words: We would _always_ store the protocol state on disk, and only keep the most recently used parts of it in memory as a cache.
+
+That would also make the unit hash file obsolete, since even after a restart, we would know what units we produced.
+
+The drawback is that it would make all the Highway algorithms — fork choice, finality detection, … — async (or even worse: blocking on I/O). It's not clear how that would fit in with the current structure of the code. And depending on the access patterns, which are hard to predict and can be influenced by an attacker, the algorithms' performance could change by orders of magnitude.
+
+
+## Prior art
+
+[prior-art]: #prior-art
+
+TBD
+
+
+## Unresolved questions
+
+[unresolved-questions]: #unresolved-questions
+
+None so far.
```

Comment:
Shouldn't we still answer the question how to incentivize validators to create signatures?

Reply:
I'd consider it beyond the scope of this CEP, which is about reducing memory usage. Incentivizing signatures is something we need to implement anyway, independent of this CEP. A proposal was also made [in Wednesday's meeting](https://casperlabs.atlassian.net/wiki/spaces/EN/pages/250282071/Consensus+design+meetings#2021-02-24):

* There is no _additional_ reward for finality signatures; instead we make the regular block reward conditional on them.
* Proposed blocks (“proto blocks”) can depend on finality signatures.
* Instead of at the end of era 5, you receive your block rewards for era 5 only in the first switch block that depends (directly or indirectly) on your finality signature on the switch block of era 5 (which in turn depends on the signatures on all its ancestors).